### PR TITLE
support first class modules

### PR DESCRIPTION
**Pull request by aspeddro**
Documentation: https://github.com/rescript-association/rescript-lang.org/issues/155

Example: https://alexfedoseev.com/blog/post/cool-things-you-can-do-with-first-class-modules-in-rescript-react and https://forum.rescript-lang.org/t/syntax-for-first-class-modules/980

Reanalyze examples:
- https://github.com/rescript-association/reanalyze/blob/master/examples/deadcode/src/FirstClassModules.res
- https://github.com/rescript-association/reanalyze/blob/master/examples/deadcode/src/TestFirstClassModules.res

### CODE REVIEW

File: queries/highlights.scm [hunk 1]
```diff
@@ -75,6 +75,8 @@
 
 (function parameter: (value_identifier) @parameter)
 
+("unpack") @function.builtin
```
Comment (aspeddro) [inline queries/highlights.scm:78]: > `unpack` is an 'unreserved keyword' (so it doesn’t collide with binding names etc).
https://forum.rescript-lang.org/t/syntax-for-first-class-modules/980/6?u=aspeddro

I'm not sure whether to highlight as `keyword`, `function` or `function.builtin`

File: grammar.js [hunk 2]
```diff
@@ -700,8 +715,17 @@ module.exports = grammar({
...
+      'module', '(',
+      choice(
+        seq(
+          field('name', choice($.module_identifier, $.module_identifier_path)),
+          optional(field('signature', seq(':', $.module_identifier))),
+          optional(field('definition', seq('(', $.module_unpack, ')')))
+        ),
+        seq(field('definition', $.block), ':', field('signature', $.module_identifier))
+      ),
+      ')'
```
Comment (nkrkv) [inline grammar.js:728]: This one looks bulky. I suggest splitting it by extracting two root seq choices into two separate rules and referring them here.

I also pretty sure that for a “signature” one can use not just `$.module_identifier` but also `$.module_identifier_path` or even a `$.module_expression`.

And I can’t imagine an example when the “definition” is used. I’ve tried to comment this line out and saw no tests failing.

File: grammar.js [hunk 3]
```diff
@@ -770,7 +794,7 @@ module.exports = grammar({
 
     type_parameter: $ => seq(
       'type',
-      $.type_identifier,
+      repeat($.type_identifier),
```
Comment (nkrkv) [inline grammar.js:797]: What’s the purpose of this `repeat`?

File: queries/highlights.scm (hunk 1 above)
Comment (nkrkv) [reply queries/highlights.scm:78]: I have no strong preference here, but I think it is a keyword.

It’s not a function you can compose, pass around, etc. It acts more like the `module(...)` but in contrast to it, the word “unpack” can also be used as a variable name in other contexts. Thanks to Tree-sitter it won’t be highlighted in other contexts, and when we’re indeed unpacking an FCM it behaves like a `module` keyword and should be highlighted similarly.

**Review by nkrkv — CHANGES_REQUESTED** [review]
Great work :fire: Thank you.

I have a few questions and thoughts though.

File: grammar.js [hunk 4]
```diff
@@ -166,7 +166,7 @@ module.exports = grammar({
       'module',
       optional('rec'),
       optional('type'),
-      field('name', $.module_identifier),
+      field('name', choice($.module_identifier, $._type_identifier)),
```
[the remark this answers is not in the document]
Comment (aspeddro) [reply grammar.js:169]: Here an valid example: https://github.com/shakacode/rescript-first-class-modules-usage

The File [Loadable.res](https://github.com/shakacode/rescript-first-class-modules-usage/blob/f6112377c4c634dde916af4c7bf5cd8c3371d925/src/Loadable.res)

```res
module type Component = {
  module type t
  let loader: unit => Promise.t<module(t)>
}
```

Maybe create a rule to match only `module type <type_identifier>`?

File: grammar.js (hunk 3 above)
Comment (aspeddro) [reply grammar.js:797]: [Renalyze examples](https://github.com/rescript-association/reanalyze/blob/a3f9bdedf0431f7dbc3422ad8e9f12df155192e1/examples/deadcode/src/TestFirstClassModules.res#L26-L30)

```res
@genType
let convertFirstClassModuleWithTypeEquations = (
  type o i,
      // ^ here
  x: module(MT with type Inner.inner = i and type outer = o),
) => x
```

File: grammar.js [hunk 5]
```diff
@@ -182,8 +182,11 @@ module.exports = grammar({
       $.module_expression,
       $.functor,
       $.extension_expression,
+      $.module_unpack,
```
[the remark this answers is not in the document]
Comment (aspeddro) [reply grammar.js]: Done 812b1ab42ba34e1cd85ad2d47224bbd39b572b9b

File: grammar.js (hunk 2 above)
Comment (aspeddro) [reply grammar.js]: Done 812b1ab42ba34e1cd85ad2d47224bbd39b572b9b

File: queries/highlights.scm (hunk 1 above)
Comment (aspeddro) [reply queries/highlights.scm]: Done 1dcfd0c03b7f82eb49fc4a43b29a153d0b82f189

File: grammar.js (hunk 3 above)
Comment (aspeddro) [reply grammar.js]: Changed to `repeat1` e0522d33d4dcc5426edaa05493e0c14b7e8ca141

File: grammar.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nkrkv) [reply grammar.js:169]: :open_mouth: I see

> Maybe create a rule to match only `module ?rec type <type_identifier>`?

The philosophy behind Tree-sitter is being as relaxed as you can until it causes some ambiguities, conflicts, etc. I think the strictness is not required here, even we know that `t` is not a valid name for a module implementation.

**Comment by nkrkv** [comment]
Looks cool now. Thank you!

**Review by nkrkv — APPROVED** [review]

**Pull request merged**